cl_dll/hud_spectator.cpp

```diff
@@ -990,10 +990,82 @@ void CHudSpectator::FindPlayer(const char *name)
 	// if we are NOT in HLTV mode, spectator targets are set on server
 	if ( !gEngfuncs.IsSpectateOnly() )
 	{
+		/*
+		 * HACK HACK HACK: AG server DLL doesn't let us "follow" a player with a specific name, but only let's us
+		 * use "follownext". This code uses follownext several times until it gets to the specific player
+		 * that the user clicked in the VGUI menu.
+		 *
+		 * Cannot use while(g_iUser2.name == current.name) because g_iUser doesn't get updated fast enough
+		 * (it's a server msg after all) which results in an infinite loop.
+		 *
+		 * If the user picks a player that has an ID that's higher than the one we're currently spectating (g_iUser2)
+		 * we step forward (reverse = 0)
+		 * If the user picks a player that has an ID that's lower than the one we're currently spectatng (g_iUser2)
+		 * we step backwards (reverse = 1)
+		 *
+		 */
+
+		std::vector<std::string> player_names;
+		std::vector<std::string> player_names_backwards;
+
+		cl_entity_t * pEnt = NULL;
+		gViewPort->GetAllPlayersInfo();
+		for (int i = 1; i < MAX_PLAYERS; i++ )
+		{
+			pEnt = gEngfuncs.GetEntityByIndex(i);
+
+			if (!gHUD.m_Spectator.IsActivePlayer(pEnt))
+				continue;
+
+			// Our user is already spectating this player, let's bail out
+			if (!strcmp(g_PlayerInfoList[i].name, name) && i == g_iUser2)
+				return;
+
+			// The current (i) player's id is higher than the one the user is currently spectating, let's save him to the forward gang
+			if ( i > g_iUser2 )
+				player_names.push_back(g_PlayerInfoList[i].name);
+
+			// The current (i) player's id is lower than the one the user is currently spectating, let's save him to the backward gang
+			else if ( i < g_iUser2 )
+				player_names_backwards.insert(player_names_backwards.begin(), g_PlayerInfoList[i].name);
+		}
+
+		int reverse = -1; // -1 so we can tell if we found the player going forwards or backwards
+		int steps = 1; // We start at 1 because the player we are currently spectating is NOT present in the list
+
+		for ( auto &player_name : player_names )
+		{
+			if(player_name == name)
+			{
+				reverse = 0; // set reverse to false, we're gonna go forward to get to our target player
+				break;
+			}
+			else
+				steps += 1;
+		}
+
+		if(reverse == -1)
+		{
+			steps = 1; // reset the steps we got from forwards
+			for ( auto &player_name_backwards : player_names_backwards )
+			{
+				if(player_name_backwards == name)
+				{
+					reverse = 1; // set reverse to true, we're gonna go backwards to get to our target player
+					break;
+				}
+				else
+					steps += 1;
+			}
+		}
+
 		char cmdstring[32];
-		// forward command to server
-		sprintf(cmdstring,"follow %s",name);
-		gEngfuncs.pfnServerCmd(cmdstring);
+		for(int i=0; i < steps; i++)
+		{
+			// forward command to server
+			sprintf(cmdstring,"follownext %i", reverse);
```

Comment:
I wonder if you could chain server cmds together with a `;` like you can do clientside. If not, is it possible that this can kick you for flooding or something?

Reply:
I just tested: you can't. The server will think that `test1;test2;test3` is one long command and not three separate ones. But I don't know if it can cause the client to be kicked.

Reply:
Hello. If you mean checks for flooding that are implemented in the AG server dll, then I don't think that would happen, this command isn't checked for flooding as opposed to e.g. "spectate", the difference can can be seen here:
https://github.com/martinwebrant/agmod/blob/master/src/dlls/agclient.cpp#L102-L125

If you meant a server code unrelated to AG (just in HL in general), I don't think they implement an anti-flood like that.
Just did a quick test with:
```
		for(int i=0; i < 5000; i++)
		{
				_snprintf( mrdka, sizeof( mrdka ), "i = %f\n", gEngfuncs.GetClientTime());
				gEngfuncs.pfnConsolePrint( mrdka );
				// forward command to server
				sprintf(cmdstring,"follownext %i", reverse);
				gEngfuncs.pfnServerCmd(cmdstring);
		}
```
Which is more than it's ever gonna call, seeing as there's only 32 players max. The client is still on the server (no kick or anything), the change gets noticed by the server, I think it drops the last few ones? But it doesn't kick me (local HLDS), the only thing that points to something being "wrong" is the HLDS log that says:
```
NET_QueuePacket:  Oversize packet from 192.168.0.59:27005
NET_QueuePacket:  Oversize packet from 192.168.0.59:27005
NET_QueuePacket:  Oversize packet from 192.168.0.59:27005
NET_QueuePacket:  Oversize packet from 192.168.0.59:27005
NET_QueuePacket:  Oversize packet from 192.168.0.59:27005
NET_QueuePacket:  Oversize packet from 192.168.0.59:27005
```
And the client with developer 1 says:
```
Netchan_Transmit:  Unreliable would overfow, ignoring
```
Which I'd assume means that it never hits the server with the ones that are already overflowed? You probably know better, but anyway, I think the chaining doesn't matter as it looks like the client (and server also?) just buffer the commands themselves. :)
 
Anyhow, this shouldn't be a problem for AG as maxplayers isn't 5000, but 32? Tested with a sane 32 and I end up spectating the same person correctly and everything all 32 commands are sent (I assume so, 
```
		for(int i=0; i < 32; i++)
		{
				_snprintf( mrdka, sizeof( mrdka ), "i = %f\n", gEngfuncs.GetClientTime());
				gEngfuncs.pfnConsolePrint( mrdka );
				// forward command to server
				sprintf(cmdstring,"follownext %i", i % 2);
				gEngfuncs.pfnServerCmd(cmdstring);
		}
```
I added that ^ and clicked the fastest way I could in the panel and it seemed to send it all and accept it server side. I also tried it on a random server with 4 players playing and the server was fine having me there. :)

I'd say we should be fine, although my test isn't very scientific.